static-alloc/src/unsync/node.rs

```diff
@@ -0,0 +1,206 @@
+use core::{
+    alloc::{Layout, LayoutErr},
+    cell::Cell,
+    mem::{self, MaybeUninit},
+    ptr::{self, NonNull},
+};
+
+use alloc::alloc::{alloc_zeroed, dealloc};
+
+use super::allocation::Allocation;
+
+pub(crate) type Link = Option<NonNull<Node>>;
+
+/// A Node represents an allocation block
+/// in which any type can be allocated.
+#[repr(C)]
+pub struct Node {
+    /// A pointer to the next node within the list.
+    /// This is wrapped in a Cell, so we can modify
+    /// this field with just an &self reference.
+    next: Cell<Link>,
+
+    /// An index into the data field. This index
+    /// will always be an index to an element
+    /// that has not been allocated into.
+    /// Again this is wrapped in a Cell,
+    /// to allow modification with just a
+    /// &self reference.
+    index: Cell<usize>,
+
+    /// The data slice of a node. This slice
+    /// may be of any arbitrary size. We use
+    /// a Cell<MaybeUninit> to allow modification
+    /// trough a &self reference, and to allow
+    /// writing uninit padding bytes.
+    data: [Cell<MaybeUninit<u8>>],
+}
+
+impl Node {
+    /// Returns the layout for the `header` of a `Node`.
+    /// The definition of `header` in this case is all the
+    /// fields that come **before** the `data` field.
+    /// If any of the fields of a Node are modified,
+    /// this function likely has to be modified too.
+    fn header_layout() -> Result<Layout, LayoutErr> {
+        Layout::new::<Cell<Link>>()
+            .extend(Layout::new::<Cell<usize>>())
+            .map(|layout| layout.0)
+    }
+
+    /// Returns the layout for an array with the size of `size`
+    fn data_layout(size: usize) -> Result<Layout, LayoutErr> {
+        Layout::new::<Cell<MaybeUninit<u8>>>()
+            .repeat(size)
+            .map(|layout| layout.0)
+    }
+
+    /// Returns a layout for a Node where the length of the data field is `size`.
+    /// This relies on the two functions defined above.
+    pub(crate) fn layout_from_size(size: usize) -> Result<Layout, LayoutErr> {
+        let layout = Self::header_layout()?.extend(Self::data_layout(size)?)?.0;
+        Ok(layout.pad_to_align())
+    }
+}
+
+/// An `AllocFailure` represents failed allocation **of** a Node.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct AllocFailure {
+    /// The size that could not be allocated
+    size: usize,
+}
+
+impl AllocFailure {
+    fn new(size: usize) -> Self {
+        Self { size }
+    }
+}
+
+/// An `AllocationError` represents a failure during the process
+/// of allocating a Node.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub enum AllocationError {
```

Comment:
I'd turn the error into an opaque type instead of an enum to minimize the exposed interface, as `LayoutErr` does for example. When some future use case requires knowing the exact cause then it would still be possible to provide an accessor with that representation.

Reply:
Also, the different variants don't really provide a useful ordering relationship so I wouldn't derive `Ord` or `PartialOrd` here.